concurrency-primer.tex

```diff
@@ -515,30 +579,235 @@ \section{Atomic operations as building blocks}
+Then the deadlock occurs.
+Therefore, we can see that deadlock happens when different threads lock in incompatible orders, 
+which leads to the system becoming immobilized as threads perpetually wait on each other. 
+
+Additionally, in \secref{shared-resources}, 
+we can see another problem with the lock: its scalability is limited.
+
+After understanding the issue that blocking mechanisms are prone to, 
+we try to achieve synchronization between threads without lock. 
+Consider the program below: if there is only a single thread, execute these operations as follows:
+
+\begin{cppcode}
+while (x == 0)
+{
+    x = 1 - x;
+}
+\end{cppcode}
+
+When executed by a single thread, these operations complete within a finite time. 
+However, with two threads executing concurrently, 
+if one thread executes \cpp|x = 1 - x| and the other thread executes \cpp|x = 1 - x| subsequently, 
+then the value of x will always be 0, which will lead to a livelock. 
+Therefore, even without any locks in concurrent threads, 
+we still cannot guarantee that the overall system can make progress toward achieving the programmer's goals.
+
+Consequently, we should not focus on comparing which communication tools or synchronization mechanisms are better, 
+but rather on exploring how to effectively use these tools in a given scenario to facilitate smooth communication between threads and achieve the programmer's goals.
+
+\section{Lock free}
+In \secref{concurrency-tool}, we explored different mechanisms based on the characteristics of concurrency tools, 
+as described in \secref{atomicity} and \secref{rmw}.
+In this section, we need to explore which strategies can help programmers to design a concurrency program 
+that allows concurrent threads to collectively ensure progress in the overall system while also improving scalability, 
+which is the initial goal of designing a concurrency program.
+First of all, we must figure out the scope of our problem.
+Understanding the relationship between the progress of each thread and the progress of the entire system is necessary.
+
+\subsection{Type of progress}
+When we consider the scenario where many concurrent threads collaborate and each thread is divided into many operations, 
+
+\textbf{Wait-Free} Every operation in every thread will be completed within a limited time. 
+This also implies that each operation contributes to the overall progress of the system.
+
+\textbf{Lock-Free} At any given moment, among all operations in every thread, 
+at least one operation contributes to the overall progress of the system. 
+However, it does not guarantee that starvation will not occur.
+
+\textbf{Obstruction-Free} At any given time, if there is only a single thread operating without interference from other threads, 
+its instructions can be completed within a finite time. However, when threads are working concurrently, 
+it does not guarantee progress.
+~\\
```

Comment:
Remove this.

Reply:
@jserv Do you want to remove any mention of "Obstruction-Free" later on? Or should I simply refrain from using the term "Obstruction-Free" and instead explain that it could result in the entire system not making any progress?

Reply:
> Do you want to remove any mention of "Obstruction-Free" later on?

No, I meant the removal of the line containing `~\\`.